### PR TITLE
[FEATURE] Master/slave connection support.

### PR DESCRIPTION
**Pull request by Zylius**
### Changes proposed in this pull request:
- Add support for master slave configuration.

This PR doesn't have breaking changes so long as users don't have **random** _read_ or _write_ keys in their connection configurations.

Example master/slave connection configuration:

```php
[
            'driver'    => 'mysql',
            'host'      => 'localhost',
            'port'      => '3306',
            'database'  => 'test',
            'username'  => 'homestead',
            'password'  => 'secret',
            'write'     => [
                'port'      => 3307,
                'user'      => 'homestead1',
                'password'  => 'secret1',
            ],
            'read' => [
                [
                    'port'     => 3308,
                    'database' => 'test2',
                ],
                [
                    'host' => 'localhost2',
                    'port' => 3309
                ],
            ],
        ];
```

Settings can be specified different for each read/write configuration or you can just specify your base settings and override them in a specific read/write configuration (as in example).

### CODE REVIEW

**Comment by patrickbrouwers** [comment]
Can you add some more information to the PR?
E.g. does it have breaking changes for using the current state of the config file?

File: src/EntityManagerFactory.php [hunk 1]
```diff
@@ -441,4 +445,44 @@ protected function registerMappingTypes(array $settings = [], EntityManagerInter
...
+            throw new \InvalidArgumentException("Parameter 'read' must be set for read/write config.");
+        }
+
+        $slaves = $driverConfig['read'];
+
+        if (!is_array($slaves) || in_array(false, array_map('is_array', $slaves))) {
+            throw new \InvalidArgumentException("Parameter 'read' must be an array containing multiple arrays.");
+        }
+
+        if (($key = array_search(0, array_map('count', $slaves))) && $key !== false) {
```
Comment (guiwoda) [inline src/EntityManagerFactory.php]: isn't `$key !== false` redundant here? if its `false`, first condition fails.
Search could return position 0, which would be false-_ish_ here as well, right?

File: tests/Configuration/Connections/MasterSlaveConnectionTest.php [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+     *
+     * @param array $resolvedBaseSettings
+     * @param array $settings
+     * @param $expectedOutput
+     *
+     * @dataProvider getMasterSlaveConnectionData
+     */
+    public function testMasterSlaveConnection(array $resolvedBaseSettings, array $settings, array $expectedOutput)
+    {
+        $this->assertEquals(
```
Comment (guiwoda) [inline tests/Configuration/Connections/MasterSlaveConnectionTest.php]: I like how you used dataProvider to reuse this case multiple times, but I find it a bit hard to maintain in this current form.
I wouldn't know where to modify each expectation for each setting.

Maybe refactor each case into its own method, then do something like:

```
public function getMasterSlaveConnectionData()
{
    return [
        $this->dummyBaseSettings(), $this->dummyConfig(), $this->dummyExpectations(),
        $this->oracleBaseSettings(), $this->oracleConfig(), $this->oracleExpectations(),
}
```

Or something along those lines?

File: tests/EntityManagerFactoryTest.php [hunk 3]
```diff
@@ -22,6 +22,7 @@
 use LaravelDoctrine\ORM\EntityManagerFactory;
 use LaravelDoctrine\ORM\Loggers\Logger;
 use LaravelDoctrine\ORM\Resolvers\EntityListenerResolver as LaravelDoctrineEntityListenerResolver;
+
```
Comment (guiwoda) [inline tests/EntityManagerFactoryTest.php]: Remove empty line between use statements.

File: tests/EntityManagerFactoryTest.php [hunk 4]
```diff
@@ -692,6 +701,146 @@ protected function enableLaravelNamingStrategy()
         $this->configuration->shouldReceive('setNamingStrategy')->once()->with($strategy);
     }
 
+    /**
+     * Data provider for testMasterSlaveConnection.
+     *
+     * @return array
+     */
+    public function getTestMasterSlaveConnectionData()
```
Comment (guiwoda) [inline tests/EntityManagerFactoryTest.php:709]: Same as comment :point_up:

**Review by guiwoda — CHANGES_REQUESTED** [review]

File: src/EntityManagerFactory.php (hunk 1 above)
Comment (maxbrokman) [reply src/EntityManagerFactory.php]: yeah this seem like it's trying to be too clever by half and will fail if the first config (position 0) is empty, poc here https://3v4l.org/N8gnk There doesn't seem to be any tests for this?

A lot of this could be replaced with

```
array_map(function($config){
    if (!is_array($config)) {
        //...
    }

    if (!count($config)) {
          //...
    }
}, $slaves);
```

only one loop, probably less buggy.

File: tests/EntityManagerFactoryTest.php [hunk 5]
```diff
@@ -692,6 +701,146 @@ protected function enableLaravelNamingStrategy()
...
+        ];
+
+        //Case #4. 'read' has empty entries
+        $inputConfig           = $dummyInputConfig;
+        $inputConfig['read'][] = [];
+
+        $out[] = [
+            $inputConfig,
+            \InvalidArgumentException::class,
+            "Parameter 'read' config no. 2 is empty."
```
Comment (maxbrokman) [inline tests/EntityManagerFactoryTest.php:753]: consider adding a test for the bug when read config 0 is empty

**Review by maxbrokman — COMMENTED** [review]
Is a duplicate of #157 ?

File: src/EntityManagerFactory.php (hunk 1 above)
Comment (Zylius) [reply src/EntityManagerFactory.php]: That's correct, it's faulty. Will remove the second check and add a test case for it.

**Comment by Zylius** [comment]
Updated it according to comments. I'm fully aware this is a duplicate of two other PRs but hopefully since there are so many PRs for this issue we'll get one in laravel doctrine and this PR will be the one getting merged.

**Review by guiwoda — APPROVED** [review]

**Comment by patrickbrouwers** [comment]
Can you add an example config to `/config`. This doesn't have to get published, but would be a good reference point for anybody wanting to use master-slave. Also a PR to our Docs repo would be very helpful.

**Comment by Zylius** [comment]
I don't know why you guys want to add sample documentation for laravels configuration file in your ORM repo but I've done it. Also, you can find actual documentation for this on the [laravel web site](https://laravel.com/docs/5.4/database#read-and-write-connections).

**Comment by Zylius** [comment]
And here is the docs PR https://github.com/laravel-doctrine/docs/pull/67.

**Comment by kpicaza** [comment]
Sorry, this pull request will be merged in future versions of package? 

It can be very powerful feature for most of big projects.

Thanks

**Comment by patrickbrouwers** [comment]
@kpicaza yes, if it passes my functional tests I'll merge it.

**Comment by kpicaza** [comment]
Can i help you on this issue improving tests or doing something that is needed?

**Comment by Zylius** [comment]
So are you guys actually planning to merge this or just ignore it for the next six months and close it like the other PRs?

I don't really care, but it seems like the community really wants it, and it doesn't look like anything is failing because of this new feature.

**Comment by guiwoda** [comment]
This PR was opened a month ago, not six. And we've shown interest in it. 
Mind your attitude, please. We appreciate your help, but just as you, we have other responsibilities outside open source. This will be reviewed and merged when we can spare the time to do so, in the meantime, people can force their dependencies to the fork if they really need it, or implement master / slave outside app, which can also be done.

**Comment by Zylius** [comment]
There must've been some misunderstanding, I was talking about the older PRs being open for six months and being closed.
Yeah, I've done the master/slave inside app thing. Just trying to help out the community.
Anyways, I appreciate your efforts in trying to make laravel a better framework!

**Comment by guiwoda** [comment]
I think you may be confusing projects: we don't have any history of closing six-month-old PRs.

I've fixed Travis configuration, now 1.3 is running (and passing) all tests. Please rebase your fork and check that tests also pass in it.

**Comment by Zylius** [comment]
Done ;)

File: config/read_write_database_sample.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    | Default Database Connection Name
+    |--------------------------------------------------------------------------
+    |
+    | Here you may specify which of the database connections below you wish
+    | to use as your default connection for all database work. Of course
+    | you may use many connections at once using the Database library.
+    |
+    */
+
+    'default' => env('DB_CONNECTION', 'auth'),
```
Comment (guiwoda) [inline config/read_write_database_sample.php]: Having an invalid default here may be confusing for users.

File: config/read_write_database_sample.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
+<?php
```
Comment (guiwoda) [inline config/read_write_database_sample.php]: Do we want a sample in code? I believe having updated docs with sample in docs will have better visibility.
Also, if syntax matches what Laravel documents, then we don't really need this at all.

File: src/EntityManagerFactory.php [hunk 8]
```diff
@@ -109,6 +110,10 @@ public function create(array $settings = [])
             $driver
         );
 
+        if ($this->isMasterSlaveConfigured($driver) && $this->hasValidMasterSlaveConfig($driver)) {
```
Comment (guiwoda) [inline src/EntityManagerFactory.php]: Refactor into:

```php
if ($this->isMasterSlaveConfigured()) {
    $this->validateMasterSlaveConfig($driver);

    $connection = (new ...);
}
```

File: src/EntityManagerFactory.php [hunk 9]
```diff
@@ -440,4 +445,44 @@ protected function registerMappingTypes(array $settings = [], EntityManagerInter
...
+
+        if (!is_array($slaves) || in_array(false, array_map('is_array', $slaves))) {
+            throw new \InvalidArgumentException("Parameter 'read' must be an array containing multiple arrays.");
+        }
+
+        if (($key = array_search(0, array_map('count', $slaves))) !== false) {
+            throw new \InvalidArgumentException("Parameter 'read' config no. {$key} is empty.");
+        }
+
+        return true;
```
Comment (guiwoda) [inline src/EntityManagerFactory.php]: By suggested refactor, this method does not need a return value

**Review by guiwoda — CHANGES_REQUESTED** [review]
Some more suggestions, I think it looks pretty good overall. Would merge as soon as this is ready!

File: config/read_write_database_sample.php (hunk 7 above)
Comment (guiwoda) [reply config/read_write_database_sample.php]: This question is meant for @patrickbrouwers actually, as he suggested adding it :smile:

Comment (patrickbrouwers) [reply config/read_write_database_sample.php]: Yes can be removed, I didn't understand at first it was the same syntax as the laravel config.

**Comment by Zylius** [comment]
Done.

**Review by guiwoda — APPROVED** [review]

**Comment by guiwoda** [comment]
I'm :+1: with this!
Shoud we merge, @patrickbrouwers ?

**Review by patrickbrouwers — APPROVED** [review]
